ldap/ldap.go

```diff
@@ -0,0 +1,156 @@
+package ldap
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/buchgr/bazel-remote/config"
+
+	auth "github.com/abbot/go-http-auth"
+	ldap "gopkg.in/ldap.v3"
+)
+
+// Cache represents a cache of LDAP query results so that many concurrent
+// requests don't DDoS the LDAP server.
+type Cache struct {
+	*auth.BasicAuth
+	m      sync.Map
+	config *config.LDAPConfig
+}
+
+type cacheEntry struct {
+	sync.Mutex
+	// Poor man's enum; nil pointer means uninitialized
+	authed *bool
+}
+
+func New(config *config.LDAPConfig) (*Cache, error) {
+	conn, err := ldap.DialURL(config.BaseURL)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	// Test the configured bind credentials
+	if err = conn.Bind(config.BindUser, config.BindPassword); err != nil {
+		return nil, err
+	}
+	return &Cache{
+		config: config,
+		BasicAuth: &auth.BasicAuth{
+			Realm: "Bazel remote cache",
+		},
+	}, nil
+}
+
+// Either query LDAP for a result or retrieve it from the cache
+func (c *Cache) checkLdap(user, password string) bool {
+	k := [2]string{user, password}
+	v, _ := c.m.LoadOrStore(k, &cacheEntry{})
+	ce := v.(*cacheEntry)
+	ce.Lock()
+	defer ce.Unlock()
+	if ce.authed != nil {
+		return *ce.authed
+	}
+	// Not initialized; actually do the query and record the result
+	authed := c.query(user, password)
+	ce.authed = &authed
+	timeout := c.config.CacheTime
+	// Don't cache a negative result for a long time; likely wrong password
+	if !authed {
+		timeout = 5 * time.Second
```

Comment:
Out of Curiosity: Why cache a negative result at all?

Reply:
If you build with invalid auth and large -j value, I guess you don't want to make several hundred LDAP requests before the client notices and stops/cancels the remaining jobs.